### PR TITLE
Rewrite most of the shell detection stuff (do not merge)

### PR DESCRIPTION
**Pull request by uranusjr**
The goal is to fix https://github.com/pypa/pipenv/issues/915#issuecomment-387138350

### CODE REVIEW

**Comment by uranusjr** [comment]
Manually tested on cmd.exe, Powershell (pwsh.exe, but powershell.exe should work), and Bash in WSL. This needs more testers.

**Comment by uranusjr** [comment]
Some explaination.

I did some process-digging and it turns out there is not a trace of `Cmder.exe` anywhere in the process tree when `_win_utils.get_shell()` is called, so I removed it from `SHELL_NAMES`. It is not useful anyway, since we really need to know not only we’re running Cmder, but also what shell it is spawning for us.

There are quite a few variants of subshell-spawning, each requiring a different call:

* Non-bash shells on non-Windows (simplest to handle).
* Bash (not including Msys Bash, which never worked).
* Bare cmd or Powershell (including pwsh)
* cmd in Cmder
* Powershell in Cmder

The logic gets complicated pretty quickly, so I decided to refactor them into five classes (plus an intermediate for the Cmder cases). `_detect_shell` now returns an instance of `Shell` (or its subclass) for `shell` to use.

File: pipenv/patched/pew/pew.py [hunk 1]
```diff
@@ -148,79 +148,123 @@ def inve(env, command, *args, **kwargs):
...
-def fork_cmder(env, cwd):
-    shell_cmd = ['cmd']
-    cmderrc_path = r'%CMDER_ROOT%\vendor\init.bat'
-    if expandpath(cmderrc_path).exists():
-        shell_cmd += ['/k', cmderrc_path]
-    if cwd:
-        os.environ['CMDER_START'] = cwd
-    fork_shell(env, shell_cmd, cwd)
 
 def _detect_shell():
```
Comment (techalchemy) [inline pipenv/patched/pew/pew.py]: The decision tree in this function is kinda messy, wonder if we can clean this up any

**Comment by techalchemy** [comment]
Overall these changes make sense.  I think we are maintaining 2 lists of windows compatible shells, we should consolidate those.  Otherwise I think we can schedule this for 11.11

**Comment by uranusjr** [comment]
Note:

* [ ] Make sure this does not overwrite #2168.
* [ ] Generate patch.

File: pipenv/patched/pew/pew.py (hunk 1 above)
Comment (uranusjr) [reply pipenv/patched/pew/pew.py]: Not sure if this is better.

Comment (uranusjr) [reply pipenv/patched/pew/pew.py]: Rewrote this a bit. Not sure if this makes more sense.

**Comment by hstefan** [comment]
[Comment on #915](https://github.com/pypa/pipenv/issues/915#issuecomment-388134146)

**Pull request closed without merging**